File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var longestOnes = function(nums, k) {
    // what is the longest subarray that contains at most k "0"s?
    let left = 0, zeroCnt = 0, maxLength = 0;

    for (var right = 0; right < nums.length; right++) {
        if (nums[right] === 0) {
            zeroCnt++;
        }
        
        while (zeroCnt === k + 1) {
            if (nums[left] === 0) {
                zeroCnt--;
            }
            left++;
        }
        
        maxLength = Math.max(right - left + 1, maxLength);
    }
    
    return maxLength;
    
};

File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.java
class Solution {
    public int longestOnes(int[] nums, int k) {
        int i = 0;
        int j = 0;
        int curr_zero = 0;
        int curr_length = 0;
        int max_length = 0;

        while (j < nums.length) {
            if (nums[j] == 0) {
                curr_zero++;
            }
            if (curr_zero <= k) {
                curr_length++;
                max_length = Math.max(max_length, curr_length);
            } else {
                while (curr_zero > k) {
                    if (nums[i] == 0) {
                        curr_zero--;
                    }
                    i++;
                }
                curr_length = j - i + 1;
            }
            j++;
        }
    return max_length;
    }
}